File: Python_Ruby/inheritance/2.rb
class Calculator
    def initialize(v1, v2)
        @v1 = v1
        @v2 = v2
    end

    def setV1(v)
        if v.is_a?(Integer)
            @v1 = v
        else
            p 'you tried changing v1 by not integer'
        end
    end

    def add()
        return @v1 + @v2
    end

    def subtract()
        return @v1 - @v2
    end
end

class MultipleCalculator < Calculator
    def multiply()
        return @v1 * @v2
    end
end

class DivideCalculator < MultipleCalculator
    def divide()
        return @v1 / @v2
    end
end

c1 = MultipleCalculator.new(10, 10)
p 'add', c1.add()
p 'subtract', c1.subtract()
p 'multiply', c1.multiply()

c2 = DivideCalculator.new(20, 10)
p 'add', c2.add()
p 'subtract', c2.subtract()
p 'multiply', c2.multiply()
p 'divide', c2.divide()

File: Python_Ruby/inheritance/2.py
class Calculator(object):
    def __init__(self, v1, v2):
        self.v1 = v1
        self.v2 = v2

    def setV1(self, v):
        if(isinstance(v, int)):
            self.v1 = v
        else:
            print('You tried changing v1 by not integer')

    def getV1(self):
        return self.v1

    def add(self):
        return self.v1 + self.v2
    def subtract(self):
        return self.v1 - self.v2

class MultipleCalculator(Calculator):
    def multiply(self):
        return self.v1 * self.v2

class DivideCalculator(MultipleCalculator):
    def divide(self):
        return self.v1 / self.v2

c1 = MultipleCalculator(10, 10)
print('add', c1.add())
print('subtract', c1.subtract())
print('multiply', c1.multiply())

c2 = DivideCalculator(20, 10)
print('add', c2.add())
print('subtract', c2.subtract())
print('multiply', c2.multiply())
print('divide', c2.divide())